### PR TITLE
Add a smoke test in build pipeline

### LINKED ISSUE #353: TypeScript build fails due to missing moment types
# Repro steps

This only repros with the bits on the dev branch, not the released package. So first do this in the durable repo:
1. `npm install`
2. `npm run build`
3. `npm pack`

In a new folder:
1. Create a sample typescript function project
2. `npm install <path to the the tgz file created in the durable repo above>`
3. Add a durable function (honestly not sure this is necessary - just having the package in your project might be enough)
1. `npm run build`

> NOTE: The sample project cannot be in a sub folder of the durable repo. The way node.js works, it'll search all parent folders for a "node_modules" folder, so if `@types/moment` package exists in a parent folder you won't get the error.

# Expected

Build passes

# Actual

You should get build errors complaining about references to moment in a types file
```
node_modules/durable-functions/lib/src/task.d.ts:4:25 - error TS2307: Cannot find module 'moment' or its corresponding type declarations.

4 import moment = require("moment");
```

# Additional info

It would be nice if we could add some sort of test to the build pipeline that essentially replicates my repro steps from above. Until #352 is fixed, it will be easy to accidentally reference types from external packages like this

### PR DESCRIPTION
**Pull request by hossam-nasr**
Resolves #353 (per suggestion in additional info), related to #354 and #356.

I've never used Azure Pipelines (or even YAML files at all) before, so I'm not sure if this is the right or the most efficient way to do this, but at least it works 😛

## Description 

This new build job essentially replicates the repro steps in #353:
1. builds and packs `durable-functions` into a `.tgz` file
2. copies the `samples/UnitTesting` directory into a `test-app` directory, outside the durable SDK directory.
    - I chose to create the test app by copying the `samples/UnitTesting` directory, but if we make the whole directory runnable (see discussions here: https://github.com/Azure/azure-functions-durable-js/issues/334#issuecomment-1091919762 -- I've also created an issue here to track this: #356), then we could instead copy the whole samples directory:
         * I didn't want to create the test app from scratch (mostly because it seemed more complicated) and preferred to just copy from the samples.
         * This was the only sample that was a fully working function app not just a sample function
         * I discovered later there are benefits to copying our samples (see below).
3. npm installs the generated `.tgz` file from step 1 in the test app
4. runs `npm run build` on the test app 

## Validation
Before merging #354 into this branch, the build pipeline failed (as expected) with this error (see [here](https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=66438&view=logs&j=d25dc844-164d-552e-f1e9-8647caa62262&t=578a2197-efdf-58e0-879d-404438bf2588)): 
```
node_modules/durable-functions/lib/src/task.d.ts(4,25): error TS2307: Cannot find module 'moment' or its corresponding type declarations.
```

After merging, the build pipeline succeeds (see [here](https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=66439&view=logs&j=d25dc844-164d-552e-f1e9-8647caa62262))

## Benefits

Besides checking for these kinds of type exporting mistakes in the future, while testing out this new pipeline, I realized there are other kinds of errors that can also be detected by this addition! I have fixed some of them in this PR too:

- If we break a previous types contract, but is not evident from our own code. For example, all arguments to `DummyOrchestrationContext` were supposed to be optional, but I didn't realize that I broke this contract with #340. It didn't cause any errors because none of our own code relied on the fact that those arguments are optional. Meanwhile, the `samples/UnitTesting` directory inside the durable directory referenced types from the npm `durable-functions` module, not from the local build. So this was only revealed when doing this kind of test.
- If some of our sample code is not runnable (related to #356). For example, the `samples/UnitTesting` app didn't import the `mocha` module, even though it referenced it. This also wasn't detected before, because `mocha` _is_ imported by the root durable SDK directory.

### CODE REVIEW

File: azure-pipelines.yml
Comment (davidmrdavid) [inline azure-pipelines.yml:93]: Should we make the value of `pckg` more concrete? So far it just grabs any `.tgz` file. Maybe we can give it the durable functions package prefix or some known substring that is a part of the generated `tgz` name to ensure we're grabbing the right payload

File: azure-pipelines.yml
Comment (davidmrdavid) [inline azure-pipelines.yml:68]: I think we could safely rename this to `SmokeTest` as it seems to test not just types, but any error that would show up when setting up a DF app.

**Review by davidmrdavid — COMMENTED** [review]
Left some small suggestions. I'll defer to @ejizba to for the final approval here, but I'd love to hear your thought son my suggestions below.

File: azure-pipelines.yml
Comment (hossam-nasr) [reply azure-pipelines.yml]: Makes sense, I added the `durable-functions-` prefix

File: azure-pipelines.yml
Comment (hossam-nasr) [reply azure-pipelines.yml]: Yes! I knew `TestExportedTypes` wasn't the best name but I couldn't figure out what to name it. `SmokeTest` is what I was looking for!

**Review by davidmrdavid — APPROVED** [review]
LGTM!

File: samples/UnitTesting/package.json
Comment (ejizba) [inline samples/UnitTesting/package.json:21]: > I didn't want to create the test app from scratch (mostly because it seemed more complicated) and preferred to just copy from the samples.

My preference is to have a separate test app in the `test` folder. I think a sample app will always have a different purpose than a test app and so we will run into conflicting goals pretty quickly. For example, for the purposes of https://github.com/Azure/azure-functions-durable-js/issues/353, we don't want _any_ `@types/` packages listed in the package.json otherwise it could hide a missing dependency error in `durable-functions`. But for a sample app you _would_ want to have `@types/` packages because that's what all users want in real life

When I think of a "smoke test", I think of a simple hello world app. I would suggest you use VS Code or core tools to create a project and add all three of the durable function triggers to create the simplest "working" app possible. Here's a tutorial if you want to follow exactly what a user would likely do: https://docs.microsoft.com/azure/azure-functions/durable/quickstart-js-vscode (Just use TypeScript instead of JavaScript and you can skip everything after you create the projects/functions)

File: azure-pipelines.yml
Comment (ejizba) [inline azure-pipelines.yml:68]: I would prefer to include this smoke test in the above "UnitTests" job (You can rename it to just "Test" if you want). Then you get the benefit of testing the whole matrix and it simplifies the yaml because it gets rid of some duplicate steps

File: azure-pipelines.yml
Comment (ejizba) [inline azure-pipelines.yml:88]: Try to use the built in tasks when possible instead of a script. That way it works better on all OS's. There's examples of the npm task above and there's a copy files task with an example here: https://github.com/Azure/azure-functions-nodejs-worker/blob/v3.x/azure-pipelines/build.yml#L76

Ping me offline and I can give you some tips on authoring/testing these yaml files if you want. Azure DevOps has some UI for editing pipelines that frankly isn't that intuitive, but is a lot easier than just modifying the yaml and hoping for the best.

File: samples/UnitTesting/package.json
Comment (hossam-nasr) [reply samples/UnitTesting/package.json:21]: Created a new test app in the test folder and using that for the smoke test now

File: azure-pipelines.yml
Comment (hossam-nasr) [reply azure-pipelines.yml]: Switched to built-in tasks! It took me a while to get it to work on all platforms (see the number of commits :p) but should be good to go now!

File: azure-pipelines.yml
Comment (hossam-nasr) [reply azure-pipelines.yml]: Done

File: test/test-app/package.json
Comment (ejizba) [inline test/test-app/package.json:1]: Can you add a simple README.md to the "test-app" folder explaining what this app is used for? Honestly just pointing to https://github.com/Azure/azure-functions-durable-js/issues/353 might be enough

File: test/test-app/package.json
Comment (ejizba) [inline test/test-app/package.json:3]: > Before merging https://github.com/Azure/azure-functions-durable-js/pull/354 into this branch, the build pipeline failed (as expected) with this error (see [here](https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=66438&view=logs&j=d25dc844-164d-552e-f1e9-8647caa62262&t=578a2197-efdf-58e0-879d-404438bf2588)):

Can you verify you did this for your final changes as well? It appears the link is for the first set of changes and just wanna make sure this still causes a build failure when expected

**Review by ejizba — APPROVED** [review]
Looks good!

File: test/test-app/package.json
Comment (hossam-nasr) [reply test/test-app/package.json:1]: Done

File: test/test-app/package.json
Comment (hossam-nasr) [reply test/test-app/package.json:3]: Did this, see the `hossamnasr/build-pipeline-test-2` branch: https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=71235&view=results.

**Pull request merged**
